src/main/java/org/osc/manager/ism/api/IsmDeviceApi.java

```diff
@@ -138,80 +189,184 @@ public Void call() throws Exception {
+                Root<DeviceMemberEntity> res = memberQuery.from(DeviceMemberEntity.class);
+                memberQuery.select(res).where(memberBuilder.and(memberBuilder.equal(r.get("name"), entity.getName())));
+
+                List<DeviceMemberEntity> deviceMemberResult = IsmDeviceApi.this.em.createQuery(memberQuery)
+                        .getResultList();
+                if (!deviceMemberResult.isEmpty()) {
+                    String msg = String.format("Device Member Entity already exists " + "name: %s\n", entity.getName());
+                    LOG.error(msg);
+                    throw new IllegalArgumentException(msg);
+                }
+                entity.setParent(devices.get(0));
+                IsmDeviceApi.this.em.persist(entity);
+                return entity;
             }
-        }).getId().toString();
+        }).getId();
     }
 
 
     @Override
     public String updateDeviceMember(ManagerDeviceMemberElement deviceElement, String name, String vserverIpAddress,
             String contactIpAddress, String ipAddress, String gateway, String prefixLength) throws Exception {
-        return deviceElement.getId();
+
+        DeviceMemberEntity deviceMember = new DeviceMemberEntity();
+        deviceMember.setId(Long.parseLong(deviceElement.getId()));
+        deviceMember.setName(name);
+        return updateDeviceMember(deviceMember);
+        // TODO - Sudhir Add other information like gateway, ipAddress
+    }
+
+    public String updateDeviceMember(DeviceMemberEntity deviceElement) throws Exception {
+
+        return this.txControl.required(new Callable<DeviceMemberEntity>() {
+
+            @Override
+            public DeviceMemberEntity call() throws Exception {
+
+                CriteriaBuilder criteriaBuilder = IsmDeviceApi.this.em.getCriteriaBuilder();
+
+                CriteriaQuery<DeviceMemberEntity> query = criteriaBuilder.createQuery(DeviceMemberEntity.class);
+                Root<DeviceMemberEntity> r = query.from(DeviceMemberEntity.class);
+                query.select(r).where(criteriaBuilder.and(criteriaBuilder.equal(r.get("id"), deviceElement.getId())));
+
+                List<DeviceMemberEntity> result = IsmDeviceApi.this.em.createQuery(query).getResultList();
+                if (result.isEmpty()) {
+                    String msg = String.format("Cannot find the Device Member Entity " + "Id: %s\n",
+                            deviceElement.getId());
+                    LOG.error(msg);
+                    throw new IllegalArgumentException(msg);
+                }
+                result.get(0).setName(deviceElement.getName());
+                IsmDeviceApi.this.em.persist(result.get(0));
+                return result.get(0);
+            }
+        }).getId();
     }
 
     @Override
     public void deleteDeviceMember(String id) throws Exception {
+
+        this.txControl.required(new Callable<Void>() {
```

Comment:
Same comment for update, enforce the id of the parent as well. The caller may intend to delete device member 1 belonging to parent 1 rather than parent 2. The parent relationship should always be enforced and validated. That is why in the rest api URL you have the parent id and also have vs id in the SDK api

Reply:
Assume that the caller does not know that the device id is globally unique, but it is unique only under the parent id. That is a common pattern for rest APIs.